PBSPro/pbspro
# Add simple test in AppVeyor

#### Bug/feature Description
* Add simple test in AppVeyor

#### Affected Platform(s)
* Windows

#### Cause / Analysis / Design
* None

#### Solution Description
* Currently in AppVeyor CI we just do build but there is not test on it and currently PTL is also not supported in Windows (However it is planned to add support for Windows in PTL but it will take time to get in) so this PR is to add simple test AppVeyor which covers basic testing of following area in PBS:
  - All Server hook events (except provision)
  - All Mom hook events
  - custom resource
  - Reservation/Job with full life cycle and requesting custom resource
  - Job attribute modification after submit (aka qalter)
  - Sever, Mom and Queue attribute modification using qmgr
  - Moving job from one queue to another queue/reservation (aka qmove)
  - Routing job from one queue to another queue (aka route queue)
  - Preemption using express queue
  - Deleting running job (aka qdel)
  - Job arrays submission
* One little change I had to do in pbs_password which allows user/admin to pass user's password on command line so that pbs_password won't prompt for entering password with keeping full backward compatible of asking password
   - This change is because till now I didn't find any tool which allows us to control stdin of pbs_password. I tried expect (from tcl/tk, also expect for windows), pexpect, winpexpect, wexpect, wintty, winpty, PowerShell's Process interaction, Windows's RM API (aka winrm), pywinrm and console but none of them works on windows so I changed pbs_password to accept user password from command line only
   - I understand that passing password on command line is not good idea from security prospective but currently I have only this way of providing user password to PBS so that PBS can run job on windows through automation
  - Since this change is not good idea from security prospective, that's the reason I have not wrote external design change for this as I don't wan't to make it supported thing as it is only used for automation purpose but in non-automation environment user/admin should use how they are using it now
   - Also this change will be useful in future while porting PTL to windows in case if we don't find any tool which allows us to control stdin of pbs_password
   - In future if we find any tool which fulfills this requirement then we should remove this change and use that tool
   - One alternate approach which I tried and works fine is to open up our internal API called pbs_encrypt_pwd in IFL API and then generate SWIG-Wrapped IFL API and use it through python but this change requires little more changes in PBS and also I wasn't sure about this approach and generating SWIG-Wrapped IFL API is not straight forward in Windows like Linux so I dropped this idea and changed pbs_password
* Also I had to write simple Retry-Command in PowerShell which allows us to retry given PowerShell cmdlet or native command on failure along with controlling retry count and interval/timeout between retry (same thing like self.server.expect in PTL).
  - However AppVeyor has appveyor-retry command but it doesn't work when you have PowerShell cmdlet with same context/session + it doesn't allows us to control retry count or interval between retry
* Removed one unused file (win_configure/msi/wix/build_wxs_source.bat) which was committed by mistake and added PBS dual license in file which doesn't have it
* Also last but not least, added support to download generated MSI from AppVeyor which can be useful to developer or user/admin who doesn't have system capable of compiling PBS and generating MSI locally

#### Testing logs/output
* Test logs: [see here](https://ci.appveyor.com/project/PBSPro/pbspro/build/%23338-pfjbibbn/job/eynh8s57m7tf5vd6#L2375)
* How download artifacts looks like: [see here](https://ci.appveyor.com/project/PBSPro/pbspro/build/%23338-pfjbibbn/job/eynh8s57m7tf5vd6/artifacts)
* Manual tests of pbs_password change
  ```
  C:\Users\appveyor>pbs_password appveyor pbS@123
  Changed user appveyor's PBS password on server APPVYR-WIN
  
  C:\Users\appveyor>pbs_password
  Enter user's password:
  Verifying - Enter user's password:
  Changed user appveyor's PBS password on server APPVYR-WIN
  
  C:\Users\appveyor>pbs_password appveyor
  Enter user's password:
  Verifying - Enter user's password:
  Changed user appveyor's PBS password on server APPVYR-WIN
  
  C:\Users\appveyor>
  ```

#### NOTE
* This change will increase total build and test time in AppVeyor by ~10 min (~5 min for Release configuration and ~5 min for Debug configuration)
   - 5 min is because installing PBS MSI and running post install script on windows takes ~3-4 min and ~1-2 min for test

#### Checklist:





__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__
